File: openmole/core/org.openmole.core.implementation/src/main/java/org/openmole/core/implementation/transition/Condition.java
/*
 *  Copyright (C) 2010 Romain Reuillon
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.core.implementation.transition;


import groovy.lang.Binding;
import java.util.Collections;
import org.openmole.core.implementation.tools.GroovyShellProxyAdapter;
import org.openmole.commons.exception.InternalProcessingError;
import org.openmole.commons.exception.UserBadDataError;

import org.openmole.core.workflow.model.job.IContext;
import org.openmole.core.workflow.model.transition.ICondition;
import org.openmole.commons.aspect.caching.ChangeState;
import org.openmole.commons.aspect.caching.SoftCachable;
import org.openmole.commons.tools.groovy.GroovyProxy;


public class Condition implements ICondition {

 	String code = "true";
	
	public Condition() {
		super();
	}
	
	public Condition(String code) {
		super();
		setCode(code);
	}

	@Override
	public boolean evaluate(IContext context) throws UserBadDataError, InternalProcessingError {
            final GroovyShellProxyAdapter groovyShellProxyAdapter = getGroovyShellProxyAdapter();
            
            synchronized(groovyShellProxyAdapter) {
                return (Boolean) groovyShellProxyAdapter.execute(context);
            }           
	}

        @ChangeState
	public void setCode(String code) {
		this.code = code;
	}


        @SoftCachable
	private GroovyShellProxyAdapter getGroovyShellProxyAdapter() throws InternalProcessingError, UserBadDataError {
            GroovyShellProxyAdapter ret = new GroovyShellProxyAdapter(new GroovyProxy());
            ret.compile(code, Collections.EMPTY_LIST);
            return ret;
        }


}

File: core/core/org.openmole.core.implementation/src/main/scala/org/openmole/core/implementation/transition/Condition.scala
/*
 * Copyright (C) 2010 Romain Reuillon
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.core.implementation.transition

import org.openmole.core.model.data._
import org.openmole.core.model.transition._
import org.openmole.misc.tools.script._
import org.openmole.core.implementation.tools._

object Condition {

  def apply(_code: String) = new Condition {
    val code = _code
  }

}

trait Condition extends ICondition {

  def code: String

  @transient lazy val groovyProxy = new GroovyProxy(code, Iterable.empty) with GroovyContextAdapter

  override def evaluate(context: Context) = groovyProxy.execute(context).asInstanceOf[Boolean]

  def unary_! = Condition(s"!($code)")

}
